File: apps/backend-java/src/main/java/com/luban/backend/service/DatasourceService.java
package com.luban.backend.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.luban.backend.dto.DatasourceResponse;
import com.luban.backend.dto.DatasourceTestResult;
import com.luban.backend.entity.Datasource;
import com.luban.backend.exception.BusinessException;
import com.luban.backend.mapper.DatasourceMapper;
import com.luban.backend.mapper.SiteMapper;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.stream.Collectors;

/**
 * Datasource CRUD + connection test. Aligned with luban-backend-go
 * internal/service/datasource_service.go (same {@code type} whitelist, same
 * DATASOURCE_NAME_CONFLICT/SITE_NOT_FOUND/DATASOURCE_NOT_FOUND error codes, same
 * TestConnection return shape).
 *
 * <p>Multi-tenant isolation: {@code get}/{@code update}/{@code delete} accept a
 * {@code siteId} and scope the row at the SQL layer (see {@link DatasourceMapper}).
 * When the API caller omits {@code siteId}, the service falls back to the id-only
 * lookup (admin/internal paths) rather than rejecting the call — preserving backward
 * compatibility with the existing contract tests.
 *
 * <p>{@code testConnection} only applies to type={@code api} (HTTP GET against the
 * configured {@code url}). For {@code static} datasources there is no remote to
 * probe — we return ok=true with latency 0 immediately, matching the Go backend.
 */
@Service
public class DatasourceService {

    private static final Logger log = LoggerFactory.getLogger(DatasourceService.class);

    /** Whitelist enforced identically on both backends (plan §3). */
    static final Set<String> ALLOWED_TYPES = Set.of("static", "api");

    private static final int TEST_TIMEOUT_SECONDS = 5;

    private final DatasourceMapper datasourceMapper;
    private final SiteMapper siteMapper;
    private final ObjectMapper objectMapper = new ObjectMapper();

    public DatasourceService(DatasourceMapper datasourceMapper, SiteMapper siteMapper) {
        this.datasourceMapper = datasourceMapper;
        this.siteMapper = siteMapper;
    }

    @Transactional(readOnly = true)
    public List<DatasourceResponse> list(String siteId) {
        if (siteId == null || siteId.isBlank()) {
            throw BusinessException.invalidArgument("siteId is required");
        }
        if (siteMapper.getById(siteId) == null) throw BusinessException.siteNotFound();
        return datasourceMapper.listBySiteId(siteId).stream()
                .map(DatasourceResponse::fromEntity)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public DatasourceResponse get(String id, String siteId) {
        Datasource d = (siteId == null || siteId.isBlank())
                ? datasourceMapper.getById(id)
                : datasourceMapper.getByIdAndSiteId(id, siteId);
        if (d == null) throw BusinessException.datasourceNotFound();
        return DatasourceResponse.fromEntity(d);
    }

    @Transactional(rollbackFor = Exception.class)
    public DatasourceResponse create(String siteId, String name, String type, JsonNode config) {
        validateType(type);
        if (siteMapper.getById(siteId) == null) throw BusinessException.siteNotFound();
        Datasource ds = new Datasource();
        ds.setId(UUID.randomUUID().toString());
        ds.setSiteId(siteId);
        ds.setName(name);
        ds.setType(type);
        ds.setConfigJson(configToJson(config, true));
        Instant now = Instant.now();
        ds.setCreatedAt(now);
        ds.setUpdatedAt(now);
        try {
            datasourceMapper.insert(ds);
        } catch (DataIntegrityViolationException e) {
            if (isDuplicate(e)) {
                log.warn("datasource create rejected: name conflict (siteId={}, name={}, type={})", siteId, name, type);
                throw BusinessException.datasourceNameConflict();
            }
            throw e;
        }
        log.info("datasource created (id={}, siteId={}, name={}, type={})", ds.getId(), siteId, name, type);
        return DatasourceResponse.fromEntity(ds);
    }

    @Transactional(rollbackFor = Exception.class)
    public DatasourceResponse update(String id, String siteId, String name, String type, JsonNode config) {
        validateType(type);
        // Tenant guard: when siteId is supplied, scope the lookup; fall back to id-only
        // for admin/internal callers (preserves backward compatibility with existing tests).
        Datasource ds = (siteId == null || siteId.isBlank())
                ? datasourceMapper.getById(id)
                : datasourceMapper.getByIdAndSiteId(id, siteId);
        if (ds == null) throw BusinessException.datasourceNotFound();
        // site_id is immutable; ignore any caller-supplied siteId override.
        ds.setName(name);
        ds.setType(type);
        ds.setConfigJson(configToJson(config, true));
        ds.setUpdatedAt(Instant.now());
        try {
            int n = datasourceMapper.update(ds);
            if (n == 0) throw BusinessException.datasourceNotFound();
        } catch (DataIntegrityViolationException e) {
            if (isDuplicate(e)) {
                log.warn("datasource update rejected: name conflict (id={}, siteId={}, name={})", id, ds.getSiteId(), name);
                throw BusinessException.datasourceNameConflict();
            }
            throw e;
        }
        log.info("datasource updated (id={}, siteId={}, name={}, type={})", id, ds.getSiteId(), name, type);
        return DatasourceResponse.fromEntity(ds);
    }

    @Transactional(rollbackFor = Exception.class)
    public void delete(String id, String siteId) {
        int n = (siteId == null || siteId.isBlank())
                ? datasourceMapper.deleteById(id)
                : datasourceMapper.deleteByIdAndSiteId(id, siteId);
        if (n == 0) throw BusinessException.datasourceNotFound();
        log.info("datasource deleted (id={}, siteId={})", id, siteId);
    }

    /**
     * Probe the datasource. For {@code api} type, issues a GET to {@code config.url}
     * (headers optional) with a 5s cap; any non-2xx or IO failure becomes a
     * {@code DATASOURCE_CONNECTION_FAILED} (503). For {@code static} type returns ok
     * immediately. Latency is measured around the HTTP call.
     *
     * <p>Security: this backend call is server-to-server; SSRF protection for
     * user-driven fetches lives in the BFF (see bff proxy/fetch route). Here we only
     * validate the configured URL parses; admins are trusted to configure legit
     * targets. No {@code @Transactional} — this method makes an outbound HTTP call
     * and must not hold a DB transaction open across it.
     */
    public DatasourceTestResult testConnection(String id) {
        Datasource ds = datasourceMapper.getById(id);
        if (ds == null) throw BusinessException.datasourceNotFound();
        if ("static".equals(ds.getType())) {
            return new DatasourceTestResult(true, "static datasource: no remote to probe", 0L);
        }
        JsonNode config = parseConfig(ds.getConfigJson());
        String url = config != null ? config.path("url").asText(null) : null;
        if (url == null || url.isBlank()) {
            throw BusinessException.datasourceConnectionFailed("config.url is required for api datasource");
        }
        URI uri;
        try {
            uri = URI.create(url);
        } catch (IllegalArgumentException e) {
            throw BusinessException.datasourceConnectionFailed("invalid config.url: " + e.getMessage());
        }
        HttpRequest.Builder reqBuilder = HttpRequest.newBuilder()
                .uri(uri)
                .timeout(Duration.ofSeconds(TEST_TIMEOUT_SECONDS))
                .GET();
        JsonNode headersNode = config.path("headers");
        if (headersNode.isObject()) {
            var fields = headersNode.fields();
            while (fields.hasNext()) {
                var entry = fields.next();
                String value = entry.getValue().isTextual() ? entry.getValue().asText() : entry.getValue().toString();
                reqBuilder.header(entry.getKey(), value);
            }
        }
        long start = System.currentTimeMillis();
        HttpClient client = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(TEST_TIMEOUT_SECONDS))
                .build();
        try {
            HttpResponse<Void> resp = client.send(reqBuilder.build(), HttpResponse.BodyHandlers.discarding());
            long latency = System.currentTimeMillis() - start;
            int sc = resp.statusCode();
            if (sc >= 200 && sc < 400) {
                return new DatasourceTestResult(true, "HTTP " + sc, latency);
            }
            log.warn("datasource connection probe failed (id={}, host={}, status={})", id, uri.getHost(), sc);
            throw BusinessException.datasourceConnectionFailed("HTTP " + sc);
        } catch (BusinessException e) {
            throw e;
        } catch (Exception e) {
            log.warn("datasource connection probe failed (id={}, host={}, err={})", id, uri.getHost(), e.getMessage());
            throw BusinessException.datasourceConnectionFailed(e.getMessage() != null ? e.getMessage() : "connection failed");
        }
    }

    private void validateType(String type) {
        if (type == null || !ALLOWED_TYPES.contains(type)) {
            throw BusinessException.invalidArgument("type must be one of " + ALLOWED_TYPES);
        }
    }

    private boolean isDuplicate(DataIntegrityViolationException e) {
        String msg = e.getMessage();
        return msg != null && (msg.contains("Duplicate") || msg.contains("Unique index") || msg.contains("uk_datasources_site_name"));
    }

    /**
     * Serialize config to a JSON string for storage.
     *
     * @param strictOnFailure when {@code true} (create/update paths), a serialization
     *                        failure throws INVALID_ARGUMENT rather than silently
     *                        degrading to {@code "{}"} — callers sent this config and
     *                        must be told it's invalid. When {@code false} (internal
     *                        read paths), we degrade to {@code "{}"} with a warn log.
     */
    private String configToJson(JsonNode config, boolean strictOnFailure) {
        if (config == null) return "{}";
        try {
            return objectMapper.writeValueAsString(config);
        } catch (Exception e) {
            if (strictOnFailure) {
                log.warn("config serialization rejected (isNull={}, isMissing={})", config.isNull(), config.isMissingNode());
                throw BusinessException.invalidArgument("config is not serializable");
            }
            log.warn("config serialization failed, degrading to empty object: {}", e.getMessage());
            return "{}";
        }
    }

    private JsonNode parseConfig(String raw) {
        if (raw == null || raw.isBlank()) return null;
        try {
            return objectMapper.readTree(raw);
        } catch (Exception e) {
            return null;
        }
    }
}

File: apps/backend-go/internal/service/datasource_service.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"time"

	"github.com/google/uuid"
	"github.com/xiaoshuai1024/luban-backend-go/internal/model"
	"github.com/xiaoshuai1024/luban-backend-go/internal/repository"
)

// DatasourceService implements datasource CRUD + connection probe, mirroring the
// Java DatasourceService. Type whitelist (static|api) and the test-connection
// behavior are identical across the two backends so the bff/engine can call either
// transparently.
type DatasourceService struct {
	repo     *repository.DatasourceRepository
	siteRepo *repository.SiteRepository
	httpc    *http.Client
}

// NewDatasourceService wires the datasource repo and a shared HTTP client for
// connection probes. The client cap matches the Java TEST_TIMEOUT_SECONDS=5s.
func NewDatasourceService(repo *repository.DatasourceRepository, siteRepo *repository.SiteRepository) *DatasourceService {
	return &DatasourceService{
		repo:     repo,
		siteRepo: siteRepo,
		httpc:    &http.Client{Timeout: 5 * time.Second},
	}
}

// AllowedTypes is exported for tests / introspection and to keep the whitelist a
// single source of truth. Both backends enforce the same set (plan §3).
var AllowedTypes = map[string]struct{}{
	"static": {},
	"api":    {},
}

// List returns every datasource owned by siteID. siteID is required; missing or
// unknown site → ErrSiteNotFound (404), so callers can't list datasources for a
// site they fabricated.
func (s *DatasourceService) List(ctx context.Context, siteID string) ([]*model.Datasource, error) {
	if siteID == "" {
		return nil, ErrInvalidArgument
	}
	if _, err := s.siteRepo.Get(ctx, siteID); err != nil {
		return nil, err
	}
	return s.repo.ListBySite(ctx, siteID)
}

// Get returns a single datasource by id, scoped to siteID when non-empty (multi-
// tenant guard). When siteID is empty, falls back to the id-only lookup (admin /
// internal path) to preserve backward compatibility with the existing contract
// tests. Missing OR wrong-tenant → ErrDatasourceNotFound (404).
func (s *DatasourceService) Get(ctx context.Context, id, siteID string) (*model.Datasource, error) {
	if siteID == "" {
		return s.repo.Get(ctx, id)
	}
	return s.repo.GetBySiteID(ctx, id, siteID)
}

// Create validates the type whitelist, checks the site exists, and inserts.
// Duplicate (site_id, name) bubbles up as ErrDatasourceNameConflict (409).
func (s *DatasourceService) Create(ctx context.Context, siteID, name, dsType string, config json.RawMessage) (*model.Datasource, error) {
	if err := validateType(dsType); err != nil {
		return nil, err
	}
	if _, err := s.siteRepo.Get(ctx, siteID); err != nil {
		return nil, err
	}
	ds := &model.Datasource{
		ID:     uuid.NewString(),
		SiteID: siteID,
		Name:   name,
		Type:   dsType,
		Config: normalizeConfig(config),
	}
	if err := s.repo.Create(ctx, ds); err != nil {
		// Audit log on failure. NEVER log config.
		log.Printf("WARN datasource create rejected siteID=%s name=%s type=%s err=%v",
			siteID, name, dsType, err)
		return nil, err
	}
	log.Printf("INFO datasource created id=%s siteID=%s name=%s type=%s", ds.ID, siteID, name, dsType)
	return ds, nil
}

// Update validates type + ownership, then overwrites the mutable fields WITHOUT
// changing site_id (the tenant never moves). It first loads the existing row via
// GetBySiteID/Get to enforce ownership AND to backfill CreatedAt so the handler
// returns a complete entity (fixes the createdAt zero-value bug, review-go-api C3).
// siteID empty → admin/internal fallback (no ownership check).
// Missing datasource → ErrDatasourceNotFound (404); name conflict → 409.
// Returns the updated entity so callers don't need a second read.
func (s *DatasourceService) Update(ctx context.Context, siteID, id, name, dsType string, config json.RawMessage) (*model.Datasource, error) {
	if err := validateType(dsType); err != nil {
		return nil, err
	}
	var existing *model.Datasource
	var err error
	if siteID == "" {
		existing, err = s.repo.Get(ctx, id)
	} else {
		existing, err = s.repo.GetBySiteID(ctx, id, siteID)
	}
	if err != nil {
		return nil, err
	}
	ds := &model.Datasource{
		ID:        id,
		SiteID:    existing.SiteID, // preserve original tenant; ignore caller-supplied siteID
		Name:      name,
		Type:      dsType,
		Config:    normalizeConfig(config),
		CreatedAt: existing.CreatedAt, // backfill so the response isn't zero-valued
	}
	if err := s.repo.Update(ctx, ds); err != nil {
		// Audit log on failure. NEVER log config (may contain secrets/headers) —
		// only id/siteID/name/type + the wrapped error.
		log.Printf("WARN datasource update failed id=%s siteID=%s name=%s type=%s err=%v",
			id, ds.SiteID, name, dsType, err)
		return nil, err
	}
	return ds, nil
}

// Delete removes a datasource, scoped to siteID when non-empty (multi-tenant guard).
// siteID empty → admin/internal fallback. Missing OR wrong-tenant →
// ErrDatasourceNotFound (404).
func (s *DatasourceService) Delete(ctx context.Context, id, siteID string) error {
	if siteID == "" {
		return s.repo.Delete(ctx, id)
	}
	return s.repo.DeleteBySiteID(ctx, id, siteID)
}

// TestConnectionResult is the response shape for POST /datasources/:id/test.
// Field names match the Java DatasourceTestResult record (ok/message/latencyMs)
// so both backends return identical JSON.
type TestConnectionResult struct {
	OK        bool   `json:"ok"`
	Message   string `json:"message"`
	LatencyMs int64  `json:"latencyMs"`
}

// TestConnection probes a datasource. For static datasources there is no remote,
// so it returns ok=true with latency 0 immediately (matches Java). For api
// datasources it issues a GET against config.url (optional config.headers) with
// the shared 5s-capped client; any non-2xx or IO error becomes
// ErrDatasourceConnectionFailed (503).
//
// SSRF note: this is a server-to-server probe against an admin-configured URL.
// User-driven arbitrary fetches go through the BFF proxy/fetch route, which owns
// the SSRF allowlist. Here we only validate that config.url parses.
func (s *DatasourceService) TestConnection(ctx context.Context, id string) (*TestConnectionResult, error) {
	ds, err := s.repo.Get(ctx, id)
	if err != nil {
		return nil, err
	}
	if ds.Type == "static" {
		return &TestConnectionResult{OK: true, Message: "static datasource: no remote to probe", LatencyMs: 0}, nil
	}

	cfg := struct {
		URL     string            `json:"url"`
		Headers map[string]string `json:"headers"`
	}{}
	if err := json.Unmarshal(ds.Config, &cfg); err != nil {
		return nil, fmt.Errorf("%w: invalid config json", ErrDatasourceConnectionFailed)
	}
	if cfg.URL == "" {
		return nil, fmt.Errorf("%w: config.url is required for api datasource", ErrDatasourceConnectionFailed)
	}

	// Explicit per-call deadline so a hung upstream can't outlive the request even
	// if the caller's ctx has no deadline. 5s matches the Java TEST_TIMEOUT_SECONDS;
	// httpc.Timeout below acts as belt-and-suspenders backstop.
	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, cfg.URL, nil)
	if err != nil {
		return nil, fmt.Errorf("%w: %s", ErrDatasourceConnectionFailed, err.Error())
	}
	for k, v := range cfg.Headers {
		req.Header.Set(k, v)
	}

	start := time.Now()
	resp, err := s.httpc.Do(req)
	if err != nil {
		// Audit: log host + err; NEVER log headers (credentials may be present).
		log.Printf("WARN datasource probe io-failed id=%s host=%s err=%v", id, req.URL.Host, err)
		return nil, fmt.Errorf("%w: %s", ErrDatasourceConnectionFailed, err.Error())
	}
	defer func() { _ = resp.Body.Close() }()
	_, _ = io.Copy(io.Discard, resp.Body)
	latency := time.Since(start).Milliseconds()

	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
		log.Printf("WARN datasource probe bad-status id=%s host=%s status=%d", id, req.URL.Host, resp.StatusCode)
		return nil, fmt.Errorf("%w: HTTP %d", ErrDatasourceConnectionFailed, resp.StatusCode)
	}
	return &TestConnectionResult{OK: true, Message: fmt.Sprintf("HTTP %d", resp.StatusCode), LatencyMs: latency}, nil
}

func validateType(t string) error {
	if _, ok := AllowedTypes[t]; !ok {
		return fmt.Errorf("%w: type must be one of static, api", ErrInvalidArgument)
	}
	return nil
}

func normalizeConfig(config json.RawMessage) json.RawMessage {
	if len(config) == 0 {
		return json.RawMessage(`{}`)
	}
	return config
}
